### PR TITLE
clean up execution errors that we expect to be persisted on-chain

### PR DESCRIPTION
**Pull request by bmwill**

### CODE REVIEW

File: crates/sui-adapter-transactional-tests/tests/children/child_of_shared_object.exp [hunk 1]
```diff
@@ -24,7 +24,7 @@ task 6 'run'. lines 86-88:
 written: object(109), object(111), object(112)
 
 task 7 'run'. lines 89-89:
-Error: When an (either direct or indirect) child object of a shared object is passed as a Move argument,either the child object's type or the shared object's type must be defined in the same module as the called function. This is violated by object fake(109) (defined in module 'T1::O1'), whose ancestor fake(111) is a shared object (defined in module 'T2::O2'), and neither are defined in this module 'T1::O1'
+Error: MiscellaneousError
```
Comment (tnowacki) [inline crates/sui-adapter-transactional-tests/tests/children/child_of_shared_object.exp:27]: This kind of sucks for these tests. Is there anything we can do for displaying/getting more info out of the adapter/execution_engine for these?

File: crates/sui-types/src/error.rs [hunk 2]
```diff
@@ -417,3 +433,193 @@ impl ExecutionStateError for SuiError {
...
+            | ExecutionErrorKind::TypeError
+            | ExecutionErrorKind::CircularObjectOwnership
+            | ExecutionErrorKind::MissingObjectOwner
+            | ExecutionErrorKind::InvalidSharedChildUse
+            | ExecutionErrorKind::ModulePublishFailure
+            | ExecutionErrorKind::ModuleVerificationFailure => {
+                ExecutionFailureStatus::MiscellaneousError
+            }
+            ExecutionErrorKind::VmError => {
+                let source = if let Some(source) = self.source() {
```
Comment (tnowacki) [inline crates/sui-types/src/error.rs:540]: I have a personal crusade here, so take it or leave it :) 
I really dislike `if-let-else` as it just feels like a worse version of `match` to me. (where `if-let` is great for not having to write `match e { p => ..., _ => () }`, I don't feel like anything is gained from `if-let` with an `else`)

Hopefully `let-else` lands some day and I can stop my crusade for cases like this one at lease.

File: crates/sui-types/src/error.rs [hunk 3]
```diff
@@ -417,3 +433,193 @@ impl ExecutionStateError for SuiError {
...
+                } else {
+                    return ExecutionFailureStatus::MiscellaneousError;
+                };
+
+                if let Some(vmerror) = source.downcast_ref::<VMError>() {
+                    match (
+                        vmerror.major_status(),
+                        vmerror.sub_status(),
+                        vmerror.location(),
+                    ) {
```
Comment (tnowacki) [inline crates/sui-types/src/error.rs:551]: Should we use VMStatus here? with the keep/discard mechanism?

Not sure how we feel about people paying for bugs or not. I think forcing people to pay for bugs is fine enough, if we are consistent about it

Comment (tnowacki) [reply crates/sui-types/src/error.rs:551]: For clarity, referring to this one:
https://github.com/move-language/move/blob/main/language/move-core/types/src/vm_status.rs#L154

**Review by tnowacki — COMMENTED** [review]
Still reviewing, but some initial comments

File: crates/sui-adapter/src/adapter.rs [hunk 4]
```diff
@@ -677,12 +661,7 @@ pub fn resolve_and_type_check(
     let fdef = match fdef_opt {
         Some(fdef) => fdef,
         None => {
-            return Err(SuiError::FunctionNotFound {
-                error: format!(
-                    "Could not resolve function '{}' in module {}",
-                    function, &module_id,
-                ),
-            })
+            return Err(ExecutionErrorKind::FunctionNotFound.into());
```
Comment (tnowacki) [inline crates/sui-adapter/src/adapter.rs:664]: My worry for losing the strings in this case, is that we really don't have another way to help people debug this. The only way devs can check their transaction is really by calling this function (which could be called from outside of execution, which I think does happen somewhere)

I think getting messages back here is probably more important for this than for the verifier

Comment (tnowacki) [reply crates/sui-adapter/src/adapter.rs:664]: Is there a reason this `ExecutionError::new_with_source` won't work here?

File: crates/sui-core/src/unit_tests/authority_tests.rs [hunk 5]
```diff
@@ -419,10 +419,10 @@ async fn test_transfer_package() {
         package_object_ref,
         gas_object.compute_object_reference(),
     );
-    let result = authority_state
+    authority_state
         .handle_transaction(transfer_transaction.clone())
-        .await;
-    assert_eq!(result.unwrap_err(), SuiError::TransferUnownedError);
+        .await
+        .unwrap_err();
```
Comment (tnowacki) [inline crates/sui-core/src/unit_tests/authority_tests.rs:425]: Can we not test the error kind at this point?

File: crates/sui-core/src/unit_tests/move_integration_tests.rs [hunk 6]
```diff
@@ -402,8 +402,8 @@ async fn test_object_owning_another_object() {
     .await
     .unwrap();
     assert!(matches!(
-        effects.status.unwrap_err().1,
-        SuiError::DeleteObjectOwnedObject
+        effects.status.unwrap_err(),
+        ExecutionFailureStatus::MiscellaneousError,
```
Comment (tnowacki) [inline crates/sui-core/src/unit_tests/move_integration_tests.rs:406]: maybe a comment at least of why there is an error

File: crates/sui-types/src/error.rs (hunk 3 above)
Comment (bmwill) [reply crates/sui-types/src/error.rs:551]: I mean this bit of code is inspired by vm status but idk if we want to use it directly

File: crates/sui-adapter/src/adapter.rs (hunk 4 above)
Comment (bmwill) [reply crates/sui-adapter/src/adapter.rs:664]: no there isn't a reason, other than when i started this PR i went one route and then backtracked so i could indeed change this to preserve this information.

File: crates/sui-verifier/src/id_leak_verifier.rs [hunk 7]
```diff
@@ -67,20 +64,18 @@ fn verify_id_leak(module: &CompiledModule) -> SuiResult {
...
-                                binary_view.identifier_at(handle.name),
-                                error
-                            ),
-                        });
-                    }
-                    _ => {
-                        panic!("Unexpected error type");
-                    }
-                },
+                BlockPostcondition::Error(err) => return Err(err),
```
Comment (tnowacki) [inline crates/sui-verifier/src/id_leak_verifier.rs:67]: Whats the error here though? Do wee need that extra context? Would be good to know what we are throwing away I guess is my point

File: crates/sui-core/src/unit_tests/authority_tests.rs (hunk 5 above)
Comment (bmwill) [reply crates/sui-core/src/unit_tests/authority_tests.rs:425]: unfortunately the error type that comes out of here is a SuiError and that context is lost now.

File: crates/sui-adapter-transactional-tests/tests/children/child_of_shared_object.exp (hunk 1 above)
Comment (bmwill) [reply crates/sui-adapter-transactional-tests/tests/children/child_of_shared_object.exp]: Hopefully this is addressed now.

File: crates/sui-types/src/error.rs (hunk 3 above)
Comment (bmwill) [reply crates/sui-types/src/error.rs:551]: Today we don't have a notion of not paying if a txn makes it this far and idk if i want to open up that discussion yet :)

File: crates/sui-adapter/src/adapter.rs (hunk 4 above)
Comment (bmwill) [reply crates/sui-adapter/src/adapter.rs:664]: Should be fixed now

File: crates/sui-verifier/src/id_leak_verifier.rs (hunk 7 above)
Comment (bmwill) [reply crates/sui-verifier/src/id_leak_verifier.rs]: Should be fixed now

File: crates/sui-adapter/src/adapter.rs [hunk 8]
```diff
@@ -691,21 +670,16 @@ pub fn resolve_and_type_check(
...
     // check arity of type and value arguments
     if fhandle.type_parameters.len() != type_args.len() {
-        return Err(SuiError::InvalidFunctionSignature {
-            error: format!(
-                "Expected {:?} type arguments, but found {:?}",
-                fhandle.type_parameters.len(),
-                type_args.len()
-            ),
-        });
+        return Err(ExecutionErrorKind::InvalidFunctionSignature.into());
```
Comment (tnowacki) [inline crates/sui-adapter/src/adapter.rs:682]: Can we get the string back for this case?

File: crates/sui-adapter/src/adapter.rs [hunk 9]
```diff
@@ -720,14 +694,7 @@ pub fn resolve_and_type_check(
...
     if parameters.len() != num_args {
-        return Err(SuiError::InvalidFunctionSignature {
-            error: format!(
-                "Expected {:?} arguments calling function '{}', but found {:?}",
-                parameters.len(),
-                function,
-                num_args
-            ),
-        });
+        return Err(ExecutionErrorKind::InvalidFunctionSignature.into());
```
Comment (tnowacki) [inline crates/sui-adapter/src/adapter.rs:697]: Ditto (missing string)

File: crates/sui-adapter/src/adapter.rs [hunk 10]
```diff
@@ -746,13 +713,7 @@ pub fn resolve_and_type_check(
...
                 CallArg::Pure(arg) => {
                     if !is_primitive(view, type_args, param_type) {
-                        return Err(SuiError::TypeError {
-                            error: format!(
-                                "Non-primitive argument at index {}. If it is an object, it must \
-                                be populated by an object ID",
-                                idx
-                            ),
-                        });
+                        return Err(ExecutionError::from_kind(ExecutionErrorKind::TypeError));
```
Comment (tnowacki) [inline crates/sui-adapter/src/adapter.rs:716]: Ditto (missing string)

File: crates/sui-adapter/src/object_root_ancestor_map.rs [hunk 11]
```diff
@@ -35,14 +34,13 @@ impl ObjectRootAncestorMap {
...
                     Owner::ObjectOwner(parent_id) => {
-                        cur_owner = *direct_owner_map.get(&parent_id.into()).ok_or(
-                            SuiError::MissingObjectOwner {
-                                child_id: cur_id,
-                                parent_id: parent_id.into(),
-                            },
-                        )?;
+                        cur_owner = *direct_owner_map
+                            .get(&parent_id.into())
+                            .ok_or(ExecutionErrorKind::MissingObjectOwner)?;
```
Comment (tnowacki) [inline crates/sui-adapter/src/object_root_ancestor_map.rs:39]: Might be helpful to specify which one for the missing object owner, but not blocking

File: crates/sui-core/src/execution_engine.rs [hunk 12]
```diff
@@ -37,11 +39,11 @@ pub fn execute_transaction_to_effects<S: BackingPackageStore>(
     native_functions: &NativeFunctionTable,
     gas_status: SuiGasStatus,
     epoch: EpochId,
-) -> SuiResult<TransactionEffects> {
+) -> SuiResult<(TransactionEffects, Option<ExecutionError>)> {
```
Comment (tnowacki) [inline crates/sui-core/src/execution_engine.rs:42]: Does this ever return an `Err` now?

File: crates/sui-transactional-test-runner/src/test_adapter.rs [hunk 13]
```diff
@@ -505,7 +507,12 @@ impl<'a> SuiTestAdapter<'a> {
                 deleted: deleted_ids,
                 events,
             }),
-            ExecutionStatus::Failure { error, .. } => Err(self.render_sui_error(*error)),
+            ExecutionStatus::Failure { error, .. } => Err(anyhow::anyhow!(
+                "Transaction Effects Status: {}\nExecution Error: {}",
+                error,
+                execution_error
+                    .expect("to have an execution error if a transaction's status is a failure")
+            )),
```
Comment (tnowacki) [inline crates/sui-transactional-test-runner/src/test_adapter.rs:515]: Will have to wrap this in `self.stabilize_str` to fix the ID issue mentioned

**Review by tnowacki — CHANGES_REQUESTED** [review]
Thanks for the test fix!  but will need to re-sanitize the string, I commented on the file for how to do that

File: crates/sui-core/src/execution_engine.rs (hunk 12 above)
Comment (bmwill) [reply crates/sui-core/src/execution_engine.rs:42]: oh nope, it doesn't

File: crates/sui-types/src/error.rs [hunk 14]
```diff
@@ -417,3 +433,193 @@ impl ExecutionStateError for SuiError {
...
+    /// Circular object ownership detected
+    CircularObjectOwnership,
+    MissingObjectOwner,
+    InvalidSharedChildUse,
+
+    ModulePublishFailure,
+    ModuleVerificationFailure,
+
+    // Move Error
+    VmError,
```
[the remark this answers is not in the document]
Comment (bmwill) [reply crates/sui-types/src/error.rs:475]: see i'd argue that we should lowercase the M in VMError ;)

**Comment by bmwill** [comment]
@tnowacki ok another round!

**Review by tnowacki — APPROVED** [review]
This looks great! Thanks again for driving this!

**Pull request merged**